Fix#10219

YouJiacheng opened the pull request:
#10219

YouJiacheng committed:
Fix#10219

jakevdp commented on tests/scipy_stats_test.py:
```diff
@@ -322,6 +322,12 @@ def args_maker():
                             tol=1e-3)
     self._CompileAndCheck(lax_fun, args_maker)
 
+  def testIssue10219(self):
```
We've stopped naming tests this way because they're less descriptive than we'd like them to be. Can you do something like this instead? https://github.com/google/jax/blob/902fc0c3d2b3ec9b6034c66074984386ec35606f/tests/lax_numpy_test.py#L2081-L2082

YouJiacheng replied:
Done.

jakevdp reviewed (changes requested):
Looks great! Thanks for the fix. Just one minor comment.

jakevdp reviewed (approved):
Thanks!

jakevdp commented:
The new implementation is causing a dtype mismatch in the `CompileAndCheck` test when X64=True, which I don't entirely understand. It may be indicative of some larger type promotion bug.

jakevdp commented:
The new implementation is causing a dtype mismatch in the `CompileAndCheck` test when X64=True, which I don't entirely understand. It may be indicative of some larger type promotion bug.

YouJiacheng commented:
Would you investigate this type promotion problem?

jakevdp commented:
Ah, I see, the issue is that you use the raw input `x` in the expression `x / 2`, so when `x` is a numpy array it follows numpy-style promotion and promotes to float64. You can fix this by calling `x = jnp.asarray(x)` at the start of the function, or alternatively by calling `lax` functions directly in place of Python arithmetic operators.

YouJiacheng commented:
Ah, it is really subtle! Calling `lax` functions can be not suitable since `lax` won't do type promotion, and need a manually promoted literal `2` (or `0.5` if use `mul`).  In addition,`lax.div(1, 2) == 0` since `1` and `2` are `int32`.
I have fixed it with
```python
def logpdf(x):
  # convert to JAX before using python arithmetic op
  # ensure correct type promotion behavior
  # see https://github.com/google/jax/pull/10220#issuecomment-1095444492
  half_x = jnp.multiply(x, 0.5)
  return -2 * jnp.logaddexp(half_x, -half_x)
```

jakevdp commented:
> Calling lax functions can be not suitable

Right, you'll see elsewhere that we use `lax._const` for this reason: https://github.com/google/jax/blob/9455254b9fd456fe0ca00851f5fd4ebfdf04aa5b/jax/_src/scipy/stats/poisson.py#L26-L31

I think that would be preferred here over introducing a `jnp` dependence.

YouJiacheng commented:
Right, but
```python
from jax._src.lax.lax import _const
assert _const(1, 0.5) == 0
```

YouJiacheng commented:
Do we need a `_promote_args_inexact` here?

YouJiacheng commented:
Done.
```python
def logpdf(x):
  x = _promote_args_inexact("logistic.logpdf", x)
  two = _lax_const(x, 2)
  half_x = lax.div(x, two)
  return lax.mul(lax.neg(two), jnp.logaddexp(half_x, lax.neg(half_x)))
```

YouJiacheng commented:
I make an empty force-push to re-run lint and typecheck which seems to fail with juptext internal bug
```
Traceback (most recent call last):
  File "/home/runner/.cache/pre-commit/repobvvyoahl/py_env-python3.8/bin/jupytext", line 8, in <module>
    sys.exit(jupytext())
  File "/home/runner/.cache/pre-commit/repobvvyoahl/py_env-python3.8/lib/python3.8/site-packages/jupytext/cli.py", line [489](https://github.com/google/jax/runs/5996732829?check_suite_focus=true#step:5:489), in jupytext
    exit_code += jupytext_single_file(nb_file, args, log)
  File "/home/runner/.cache/pre-commit/repobvvyoahl/py_env-python3.8/lib/python3.8/site-packages/jupytext/cli.py", line 557, in jupytext_single_file
    notebook = read(nb_file, fmt=fmt, config=config)
  File "/home/runner/.cache/pre-commit/repobvvyoahl/py_env-python3.8/lib/python3.8/site-packages/jupytext/jupytext.py", line 409, in read
    return read(stream, as_version=as_version, fmt=fmt, config=config, **kwargs)
  File "/home/runner/.cache/pre-commit/repobvvyoahl/py_env-python3.8/lib/python3.8/site-packages/jupytext/jupytext.py", line 418, in read
    return reads(fp.read(), fmt, config=config, **kwargs)
  File "/home/runner/.cache/pre-commit/repobvvyoahl/py_env-python3.8/lib/python3.8/site-packages/jupytext/jupytext.py", line 370, in reads
    notebook = reader.reads(text, **kwargs)
  File "/home/runner/.cache/pre-commit/repobvvyoahl/py_env-python3.8/lib/python3.8/site-packages/jupytext/jupytext.py", line 178, in reads
    return new_notebook(cells=cells, metadata=metadata)
  File "/home/runner/.cache/pre-commit/repobvvyoahl/py_env-python3.8/lib/python3.8/site-packages/nbformat/v4/nbbase.py", line 164, in new_notebook
    validate(nb)
  File "/home/runner/.cache/pre-commit/repobvvyoahl/py_env-python3.8/lib/python3.8/site-packages/nbformat/v4/nbbase.py", line 39, in validate
    return validate(node, ref=ref, version=nbformat)
  File "/home/runner/.cache/pre-commit/repobvvyoahl/py_env-python3.8/lib/python3.8/site-packages/nbformat/validator.py", line 281, in validate
    raise ValidationError("Non-unique cell id '{}' detected. Corrected to '{}'.".format(cell_id, cell['id']))
jsonschema.exceptions.ValidationError: Non-unique cell id '1022fbd7' detected. Corrected to 'b1d2593d'.

Error: The process '/opt/hostedtoolcache/Python/3.8.12/x64/bin/pre-commit' failed with exit code 1
```
